rfcs/2020-09-29-4155-aws-s3-source.md

```diff
@@ -0,0 +1,187 @@
+## RFC #4155 - 2020-09-29 - AWS S3 source
+
+This RFC describes a new `aws_s3` source for ingesting events stored as objects
+in S3 via SQS bucket notifications. This will permit, for example, ingesting
+CloudTrail events, VPC flow logs, and ELB logs which can be written to S3, as
+well as custom logging pipelines writing there.
+
+## Scope
+
+This RFC will cover the implementation of an `aws_s3` source that will rely on
+SQS for bucket notifications.
+
+It will not cover the [proposed polling
+strategy](https://github.com/timberio/vector/issues/1017#issuecomment-694287111),
+but will leave space for it to be added later in a backwards compatible fashion.
+
+It will not cover parsing of events within the objects. Vector's current
+approach for this is to delegate to transforms. For example, it is likely we'll
+want to add a transform to extract CloudTrail events stored as objects in S3.
```

Comment:
It might be a feature that still needs to exist. 
I am not sure if customers would want to configure their AWS services to invoke additional costs in order to work-around a limitation in their log router.  Not that an S3 bucket of CloudTrail logs would be particularly expensive, but it's the principle of the matter.  (that said, my org is already routing CloudTrail to s3)

Reply:
For CloudTrail, specifically, I think the only options for ingestion by vector are [shipping to S3](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-create-a-trail-using-the-console-first-time.html) or [shipping to CloudWatch Logs](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/send-cloudtrail-events-to-cloudwatch-logs.html). Are you aware of any other options? [`LookupEvents`](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudtrail/lookup-events.html) is limited to 2 req/s which seems like it would make it infeasible for scraping.

I think the naive `aws_cloudwatch_logs` source (https://github.com/timberio/vector/issues/3077) is still relevant apart from CloudTrail.

We may want a transform specifically [to handle CloudTrail events in S3](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-event-reference.html), but I'm not 100% sure. Just using the `json_parser` transform seems like it might be sufficient. Again, this relates back to the idea of Vector config macros that would allow users to more easily configure ingesting CloudTrail events from S3 without needing to string together a source with relevant transforms.